pd/src/app.rs

```diff
@@ -355,22 +355,32 @@ impl App {
         let state = self.state.clone();
         async move {
             if epoch.end_height().value() == height {
+                // We've finished processing the last block of `epoch`, so we've
+                // crossed the epoch boundary, and (prev | current | next) are:
+                let prev_epoch = epoch;
+                let current_epoch = prev_epoch.next();
+                let next_epoch = current_epoch.next();
+
                 tracing::info!(
                     ?height,
-                    ?epoch,
-                    "last block of epoch, processing rate updates"
+                    ?prev_epoch,
+                    ?current_epoch,
+                    ?next_epoch,
+                    "crossed epoch boundary, processing rate updates"
                 );
                 metrics::increment_counter!("epoch");
 
                 // TODO (optimization): batch these queries
-                let current_base_rate = state.base_rate_data(epoch.index).await?;
-                let current_rates = state.rate_data(epoch.index).await?;
+                let current_base_rate = state.base_rate_data(current_epoch.index).await?;
+                let current_rates = state.rate_data(current_epoch.index).await?;
 
                 // steps (foreach validator):
                 // - get the total token supply for the validator's delegation tokens
-                // - process the updates to the token supply
-                // - call next_rates.voting_power()
-                // - persist both the voting power and the updated supply
+                // - process the updates to the token supply:
+                //   - collect all delegations occurring in previous epoch and apply them (adds to supply);
+                //   - collect all undelegations started in previous epoch and apply them (reduces supply);
+                // - feed the updated (current) token supply into current_rates.voting_power()
+                // - persist both the current voting power and the current supply
 
                 /// FIXME: set this less arbitrarily, and allow this to be set per-epoch
                 /// 3bps -> 11% return over 365 epochs, why not
```

Comment:
Is this meant to be 30,000? `3_0000` is a kind of misleading way to write that, since it looks like 3,000 at a glance.

Reply:
I's not meant to be 30,000, it's meant to be 3 basis points, and our rates are expressed in basis points of basis points (i.e., a factor of 10^(4+4)), so we want a value of 3 * 10^4 (which is 30,000, but only incidentally).